app.go

```diff
@@ -92,241 +146,132 @@ type App struct {
+func (app *App) GetServerID() string {
 	return app.server.ID
 }
 
 // GetConfig gets the pitaya config instance
-func GetConfig() *config.Config {
+func (app *App) GetConfig() *config.Config {
 	return app.config
 }
 
 // GetMetricsReporters gets registered metrics reporters
-func GetMetricsReporters() []metrics.Reporter {
+func (app *App) GetMetricsReporters() []metrics.Reporter {
 	return app.metricsReporters
 }
 
-// SetRPCServer to be used
-func SetRPCServer(s cluster.RPCServer) {
-	app.rpcServer = s
-}
-
-// SetRPCClient to be used
-func SetRPCClient(s cluster.RPCClient) {
-	app.rpcClient = s
-}
-
-// SetServiceDiscoveryClient to be used
-func SetServiceDiscoveryClient(s cluster.ServiceDiscovery) {
-	app.serviceDiscovery = s
-}
-
-// SetSerializer customize application serializer, which automatically Marshal
-// and UnMarshal handler payload
-func SetSerializer(seri serialize.Serializer) {
-	app.serializer = seri
-}
-
-// GetSerializer gets the app serializer
-func GetSerializer() serialize.Serializer {
-	return app.serializer
-}
-
 // GetServer gets the local server instance
-func GetServer() *cluster.Server {
+func (app *App) GetServer() *cluster.Server {
 	return app.server
 }
 
 // GetServerByID returns the server with the specified id
-func GetServerByID(id string) (*cluster.Server, error) {
+func (app *App) GetServerByID(id string) (*cluster.Server, error) {
 	return app.serviceDiscovery.GetServer(id)
 }
 
 // GetServersByType get all servers of type
-func GetServersByType(t string) (map[string]*cluster.Server, error) {
+func (app *App) GetServersByType(t string) (map[string]*cluster.Server, error) {
 	return app.serviceDiscovery.GetServersByType(t)
 }
 
 // GetServers get all servers
-func GetServers() []*cluster.Server {
+func (app *App) GetServers() []*cluster.Server {
 	return app.serviceDiscovery.GetServers()
 }
 
-// AddMetricsReporter to be used
-func AddMetricsReporter(mr metrics.Reporter) {
-	app.metricsReporters = append(app.metricsReporters, mr)
-}
-
-func startDefaultSD() {
-	// initialize default service discovery
-	var err error
-	app.serviceDiscovery, err = cluster.NewEtcdServiceDiscovery(
-		app.config,
-		app.server,
-		app.dieChan,
-	)
-	if err != nil {
-		logger.Log.Fatalf("error starting cluster service discovery component: %s", err.Error())
-	}
-}
-
-func startDefaultRPCServer() {
-	// initialize default rpc server
-	rpcServer, err := cluster.NewNatsRPCServer(app.config, app.server, app.metricsReporters, app.dieChan)
-	if err != nil {
-		logger.Log.Fatalf("error starting cluster rpc server component: %s", err.Error())
-	}
-	SetRPCServer(rpcServer)
-}
-
-func startDefaultRPCClient() {
-	// initialize default rpc client
-	rpcClient, err := cluster.NewNatsRPCClient(app.config, app.server, app.metricsReporters, app.dieChan)
-	if err != nil {
-		logger.Log.Fatalf("error starting cluster rpc client component: %s", err.Error())
-	}
-	SetRPCClient(rpcClient)
+// SetLogger logger setter
+func SetLogger(l logger.Logger) {
```

Comment:
Why is this one still static?

Reply:
That's a good question. I didn't make the change because it wasn't critical for my end objective, which was making pitaya more testable.
Also, I didn't want to make another shotgun surgery. Maybe we can do this in a later PR.

Reply:
Cool.